Split skip ruff field into two

krishnan-chandra opened the pull request:
Closes #20620.

Added tests for both new fields; I couldn't figure out how to get the deprecated alias `skip_ruff` to still work in tests.

I had another version of this change which kept the `skip_ruff` behavior as it is now to skip **both** linting and formatting; if that's a preferred solution, I can tack on those changes to this branch.

krishnan-chandra commented on src/python/pants/backend/python/lint/ruff/skip_field.py:
```diff
@@ -11,17 +11,31 @@
-    help = "If true, don't run ruff on this target's code."
+    help = "If true, don't run the ruff checker on this target's code."
+
+    deprecated_alias = "skip_ruff"
```
How does this feature work? When I substitute `skip_ruff` instead of `skip_ruff_check` in tests, the tests fail.

huonw replied:
Hm, good question. It seems like it sort-of works. When I run this PR via `PANTS_SOURCE`, on a repo with `skip_ruff` field on a file:

- the field is understood (i.e. no error about an unknown field, like for `skip_not_a_real_field`) ✅ 
- it doesn't do any skipping, i.e. ruff is still run ❌  (probably the same as your observation of tests failing).
- there's no deprecation warning ❌ 

Thus, maybe we instead have to have have two dedicated field classes, both the old and new, and check for them both explicitly. One might then be able to use `removal_version`/`removal_hint` for the deprecation 🤞 , or maybe one has to go all the way down to a `warn_or_error` call somewhere (not sure where).

kaos replied:
I suspect you found a bug where a target generator is not propagating a deprecated field properly. I'm investigating.. :)

krishnan-chandra replied:
> Thus, maybe we instead have to have have two dedicated field classes, both the old and new, and check for them both explicitly.

Right. I can preserve `SkipRuffField` with its old behavior as @kaos suggested here: https://github.com/pantsbuild/pants/issues/20620#issuecomment-1966087173

Then, we don't have to deprecate anything 🤷

kaos replied:
Sounds good to me. Will still fix the underlying issue though. #20627

krishnan-chandra replied:
Restored `skip_ruff` as an option and made it so that the check/format rules will skip if either `skip_ruff_check`/`skip_ruff_format` or `skip_ruff` are set.

huonw commented on src/python/pants/backend/python/lint/ruff/skip_field.py:
```diff
@@ -11,17 +11,31 @@
+
+
+class SkipRuffFormatField(BoolField):
+    alias = "skip_ruff_format"
```
It's a bit annoying to do and definitely in the realm of fine-tuning, but I wonder if it'd be better to split these into their respective backends. In particular, I believe this current arrangement means that someone only using the `...ruff.check` backend would still have a `skip_ruff_format` attached to all the Python targets, and vice versa.

krishnan-chandra replied:
I took a crack at splitting this out - main thing that makes me nervous about this is the number of files/rules moved around; I'm not sure if some rule will get omitted somewhere that causes a regression.

huonw reviewed (commented):
Nice one, thanks for taking this on!

kaos commented on src/python/pants/backend/python/lint/ruff/rules_integration_test.py:
```diff
@@ -7,16 +7,20 @@
 
 import pytest
 
+import pants.backend.python.lint.ruff.check.skip_field
+import pants.backend.python.lint.ruff.format.skip_field
```
why not import the field classes directly here, makes the lines considerably shorter later in this file..?
```suggestion
from pants.backend.python.lint.ruff.check.skip_field import SkipRuffCheckField
from pants.backend.python.lint.ruff.format.skip_field import SkipRuffFormatField
```

krishnan-chandra replied:
Sorry about this, it looks like my editor introduced this when refactoring the files. I pushed up a fix replacing this import.

huonw reviewed (approved):
Thanks!

huonw commented:
Question: is it worth cherry-picking this to 2.20?

In favour: this isn't particularly risky and is related to a "major" feature that'll be released in 2.20. Against: it's more of a new feature than a bug fix, and it's not actually changing behaviour for existing users (since `skip_ruff` keeps working just fine).

Any other reasons for or against? Thoughts?

krishnan-chandra commented:
The only other concern here is that I don't know what else might break as a result of this change - ex: the rules for the Ruff subsystem involve pulling all sources from the same directory (which has now changed as we moved check/format rules into separate directories). I tried to cover for this by including the explicit check/format rules wherever I could.

The tests didn't catch anything but I'm unsure as to what new bugs might come up that we haven't thought of.

huonw reviewed (approved):
Sweet. I'll mark this for cherry picking and merge it, and we can work out whether we want to merge the cherry pick on that PR itself.
